xarray/core/indexing.py

```diff
@@ -177,18 +206,36 @@ def convert_label_indexer(index, label, index_name='', method=None,
             if np.any(indexer < 0):
                 raise KeyError('not all values found in index %r'
                                % index_name)
-    return indexer
+    return indexer, new_index
 
 
 def remap_label_indexers(data_obj, indexers, method=None, tolerance=None):
     """Given an xarray data object and label based indexers, return a mapping
-    of equivalent location based indexers.
+    of equivalent location based indexers. Also return a mapping of pandas'
+    single index objects returned from multi-index objects.
     """
     if method is not None and not isinstance(method, str):
         raise TypeError('``method`` must be a string')
-    return dict((dim, convert_label_indexer(data_obj[dim].to_index(), label,
-                                            dim, method, tolerance))
-                for dim, label in iteritems(indexers))
+
+    pos_indexers, new_indexes = {}, {}
+    for dim, label in iteritems(indexers):
+        index = data_obj[dim].to_index()
+
+        if isinstance(index, pd.MultiIndex):
+            # set default names for multi-index unnamed levels so that
+            # we can safely rename dimension / coordinate later
```

Comment:
Shouldn't be better to move this logic to `indexing.PandasIndexAdapter`, wrap the `names` (and/or `name`) property, and return a `PandasIndexAdapter` object from `convert_label_indexer`?

This is because I worry about implicit copy or renaming.

The problem would be to set default level names that are unique across dimensions, but maybe we can pass the variable name in the `PandasIndexAdapter` constructor (as an additional `name` kwarg).